File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string expandCenter(string s, int l, int r) {
        while(l >= 0 && r < s.length() && s[l] == s[r]) {
            l--;
            r++;
        }

        return s.substr(l+1, r-l-1);
    }

    string longestPalindrome(string s) {
        int n = s.size();
        
        string max_str = s.substr(0,1);

        for(int i = 0; i < n; i++) {
            string odd = expandCenter(s, i, i);
            string even = expandCenter(s, i, i+1);

            if(odd.length() > max_str.length()) {
                max_str = odd;
            }

            if(even.length() > max_str.length()) {
                max_str = even;
            }
        }

        return max_str;
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        resIdx, resLen = 0, 0
        n = len(s)

        dp = []
        for _ in range(n):
            row = [False] * n
            dp.append(row)

        for i in range(n-1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and (j - i <= 2 or dp[i+1][j-1]):
                    dp[i][j] = True

                    if resLen < j - i + 1:
                        resIdx = i 
                        resLen = j - i + 1

        return s[resIdx : resIdx + resLen]                
        